repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Authorization.scala

```diff
@@ -0,0 +1,152 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.filters.keystonev2
+
+import javax.servlet.http.HttpServletResponse.{SC_FORBIDDEN, SC_UNAUTHORIZED}
+
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
+import org.openrepose.filters.keystonev2.AbstractKeystoneV2Filter.Reject
+import org.openrepose.filters.keystonev2.KeystoneV2Common.{Endpoint, EndpointsData, Role, ValidToken}
+import org.openrepose.filters.keystonev2.config._
+
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
+
+object KeystoneV2Authorization extends LazyLogging {
+
+  def doAuthorization(config: KeystoneV2Config, request: HttpServletRequestWrapper, validToken: ValidToken, endpoints: => Try[EndpointsData]): AuthorizationInfo = {
+    lazy val tenantToMatch: String =
+      Option(config.getTenantHandling.getValidateTenant).flatMap(validateTenantConfig =>
+        Option(validateTenantConfig.getUriExtractionRegex).flatMap(uriExtractionRegexList =>
+          uriExtractionRegexList.asScala.toStream.map(_.r).flatMap(uriExtractionRegex =>
+            request.getRequestURI match {
+              case uriExtractionRegex(tenantId, _*) => Option(tenantId)
+              case _ => None
+            }
+          ).headOption
+        )).getOrElse(throw UnparseableTenantException("Could not parse tenant from the URI"))
+
+
+    val tenantScopedRoles = getTenantScopedRoles(config.getTenantHandling.getValidateTenant, tenantToMatch, validToken.roles)
+    val userIsPreAuthed = isUserPreAuthed(config.getPreAuthorizedRoles, tenantScopedRoles)
+    val scopedRolesToken = if (userIsPreAuthed) validToken else validToken.copy(roles = tenantScopedRoles)
+    val doTenantCheck = shouldAuthorizeTenant(config.getTenantHandling.getValidateTenant, userIsPreAuthed)
+    val matchedTenant = getMatchingTenant(config.getTenantHandling.getValidateTenant, tenantToMatch, doTenantCheck, scopedRolesToken)
+    val authResult = authorizeTenant(doTenantCheck, matchedTenant) flatMap { _ =>
+      authorizeEndpoints(config.getRequireServiceEndpoint, userIsPreAuthed, endpoints)
+    }
+    authResult match {
+      case Success(_) => AuthorizationPassed(scopedRolesToken, matchedTenant)
+      case Failure(exception) => AuthorizationFailed(scopedRolesToken, matchedTenant, exception)
+    }
+  }
+
+  def handleFailures(authResult: Try[Unit.type]): Option[Reject] = {
+    authResult match {
+      case Failure(e: InvalidTenantException) => Option(Reject(SC_UNAUTHORIZED, Some(e.getMessage)))
+      case Failure(e: UnauthorizedEndpointException) => Option(Reject(SC_FORBIDDEN, Some(e.getMessage)))
+      case Failure(e: UnparseableTenantException) => Option(Reject(SC_UNAUTHORIZED, Some(e.getMessage)))
+      case _ => None
+    }
+  }
```

Comment:
Rather than making this a method, you could make this a function. It saves you the parameter and the match statement.

```
val handleFailures: (Try[Unit.type] => Option[Reject]) = {
  case Failure(e: InvalidTenantException) => Option(Reject(SC_UNAUTHORIZED, Some(e.getMessage)))
  case Failure(e: UnauthorizedEndpointException) => Option(Reject(SC_FORBIDDEN, Some(e.getMessage)))
  case Failure(e: UnparseableTenantException) => Option(Reject(SC_UNAUTHORIZED, Some(e.getMessage)))
  case _ => None
}
```

If you'd like to be a bit clearer than that, you could also give the function a type:
```
type FailureHandler = (Try[Unit.type] => Option[Reject])
val failureHandler: FailureHandler = {
  case Failure(e: InvalidTenantException) => Option(Reject(SC_UNAUTHORIZED, Some(e.getMessage)))
  case Failure(e: UnauthorizedEndpointException) => Option(Reject(SC_FORBIDDEN, Some(e.getMessage)))
  case Failure(e: UnparseableTenantException) => Option(Reject(SC_UNAUTHORIZED, Some(e.getMessage)))
  case _ => None
  }
```

It doesn't make much of a difference, but I think it's kind of neat.

Reply:
Actually, the return type should probably be:
```
type FailureHandler = PartialFuction[Try[Unit.type], Option[Reject]]
```

That way you can combine partial functions to handle failure handling the way you are doing now, as a chain of sorts.